src/Formats/SchemaInferenceUtils.cpp

```diff
@@ -307,6 +308,22 @@ namespace
         type_indexes.erase(TypeIndex::UInt64);
     }
 
+    /// if setting input_format_json_infer_variant_from_multi_type_array is true
+    /// and nested types are not equal then we convert to type variant.
+    void transformVariant(DataTypes & data_types, TypeIndexesSet & type_indexes)
+    {
+        auto variant_type = std::make_shared<DataTypeVariant>(data_types);
+        /// replace separate types with a single variant type
+        data_types.clear();
+        type_indexes.clear();
+        data_types.push_back(variant_type);
+        type_indexes.insert(TypeIndex::Variant);
+
+        /// make the second type variant as well
+        data_types.push_back(variant_type);
+        type_indexes.insert(TypeIndex::Variant);
```

Comment:
We might have any number of types, not only 2. We should iterate over `data_types` and replace each type with `Variant`